core/domain/user_services_test.py

```diff
@@ -1150,3 +1153,179 @@ def test_last_exp_edit_time_gets_updated(self):
         self.assertGreater(
             (owner_settings.last_created_an_exploration),
             previous_last_created_an_exploration)
+
+
+class UserSettingsTests(test_utils.GenericTestBase):
+
+    def setUp(self):
+        super(UserSettingsTests, self).setUp()
+        self.signup(self.OWNER_EMAIL, self.OWNER_USERNAME)
+        self.owner_id = self.get_user_id_from_email(self.OWNER_EMAIL)
+        self.owner = user_services.UserActionsInfo(self.owner_id)
+        self.signup(self.NEW_USER_EMAIL, self.NEW_USER_USERNAME)
+        self.guest_id = self.get_user_id_from_email(self.NEW_USER_EMAIL)
```

Comment:
I'm getting a bit confused about what a "guest" is. Is a guest a non-logged-in user? If so, then why do they have an email and username? If not, then what are guests allowed to do?

Reply:
Yes you are correct. Should I change this to self.user_id?

Reply:
I don't know, what is it that you actually want to do here? (What are you trying to simulate as a guest user?)

If a logged-out user, then ... well, we have lots of logged-out tests, and the way to simulate that is just to not log in in the first place.

Reply:
No I just want to test for a user which has no username. How can I do that?

Reply:
First figure out how such a user can arise in the codebase (if it does happen, it will happen during the signup flow where the user has "signed in with Google" but has not chosen a username). I would guess that you do self.get_user_id_from_email(self.NEW_USER_EMAIL)
but with no self.signup step beforehand, but I am not 100% sure.

Reply:
I tried using get_user_id_from_email without signup but that actually returns an error from [here](https://github.com/oppia/oppia/blob/31093ea610e564857f5f2c56465cb48d55381f46/core/domain/user_services.py#L276). What I found out is that signup takes us to the SignupHandler in profile.py but we can't have empty username as we will get an exception from [here](https://github.com/oppia/oppia/blob/d1d8e53bed78e7716b8af30752e7ff278bec5aa4/core/controllers/profile.py#L305). What I found out from comments in the codebase is that not registered users will have no username but for that there won't exist a UserSettings object in the first place.

So can [this line](https://github.com/oppia/oppia/blob/31093ea610e564857f5f2c56465cb48d55381f46/core/domain/user_services.py#L1013) be removed then?

Reply:
You could raise an exception if it shouldn't happen in practice?

Reply:
Yes it shouldn't I think and in that case the exception will be raised from 4 lines above: 
https://github.com/oppia/oppia/blob/31093ea610e564857f5f2c56465cb48d55381f46/core/domain/user_services.py#L1006